src/check_presence.c

```diff
@@ -0,0 +1,533 @@
+/**
+ * collectd - src/check_presence.c
+ * Copyright (C) 2019       Graham Leggett
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Graham Leggett <minfrin at sharp.fm>
+ *
+ **/
+
+/*
+ * This plugin keeps track of hosts that have been seen by collectd, and sends
+ * notifications when a new host is seen, and when a host has not been seen.
+ *
+ */
+
+#include "collectd.h"
+
+#include "plugin.h"
+#include "utils/common/common.h"
+
+#include <lmdb.h>
+
+#define PLUGIN_NAME "check_presence"
+
+#define DEFAULT_STATE_DATASTORE "check-presence"
+#define DEFAULT_HOST_TIMEOUT 20
+#define DEFAULT_THREAD_INTERVAL 2
```

Comment:
There is absolutely no need in scanning database each 2 seconds if your `HostTimeout 60`.
That is enough, say, to do such scan once per 10 seconds, then you will get notifications for hosts unreachable for 60-70seconds with average value of 65 seconds. This will reduce cpu usage 5 times.

I.e propose to made this configurable with default value of 1/2 of `HostTimeout`.

And of course, remove `thread_interval` from "avoid race conditions" code.

Reply:
CheckInterval added as recommended.